libraries/matrix/impl/src/main/kotlin/io/element/android/libraries/matrix/impl/room/member/RoomMemberListFetcher.kt

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright (c) 2024 New Vector Ltd
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.element.android.libraries.matrix.impl.room.member
+
+import io.element.android.libraries.core.coroutine.parallelMap
+import io.element.android.libraries.matrix.api.room.MatrixRoomMembersState
+import io.element.android.libraries.matrix.api.room.roomMembers
+import kotlinx.collections.immutable.toImmutableList
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.ensureActive
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+import kotlinx.coroutines.withContext
+import org.matrix.rustcomponents.sdk.RoomInterface
+import org.matrix.rustcomponents.sdk.RoomMembersIterator
+import org.matrix.rustcomponents.sdk.use
+import timber.log.Timber
+import kotlin.coroutines.cancellation.CancellationException
+
+/**
+ * This class fetches the room members for a given room in a 'paginated' way, and taking into account previous cached values.
+ */
+internal class RoomMemberListFetcher(
+    private val room: RoomInterface,
+    private val dispatcher: CoroutineDispatcher,
+    private val pageSize: Int = 1000,
+) {
+    private val updatedRoomMemberMutex = Mutex()
+    private val roomId = room.id()
+
+    private val _membersFlow = MutableStateFlow<MatrixRoomMembersState>(MatrixRoomMembersState.Unknown)
+    val membersFlow: StateFlow<MatrixRoomMembersState> = _membersFlow
+
+    /**
+     * Fetches the room members for the given room.
+     * It will emit the cached members first, and then the updated members in batches of [pageSize] items, through [membersFlow].
+     * @param withCache Whether to load the cached members first. Defaults to true.
+     */
+    @Suppress("InstanceOfCheckForException")
+    suspend fun getUpdatedRoomMembers(withCache: Boolean = true) {
```

Comment:
Weird to have a method starting with `get` that return Unit. Maybe `triggerRoomMembersUpdate` would be better?

Reply:
https://github.com/element-hq/element-x-android/pull/2274/commits/371e6c51c7fe846af848775048e8251d3b083732